ckan/logic/auth/update.py

```diff
@@ -34,10 +33,24 @@ def package_update(context, data_dict):
                 'create_unowned_dataset',
                 )) or authz.has_user_permission_for_some_org(
                 user, 'create_dataset')
+
     if not check1:
-        return {'success': False,
-                'msg': _('User %s not authorized to edit package %s') %
-                        (str(user), package.id)}
+        success = False
+        if authz.check_config_permission('allow_dataset_collaborators'):
+            # if org-level auth failed, check dataset-level auth
+            # (ie if user is a collaborator)
+            user_packages = logic.get_action(
+                'package_member_list_for_user')(
+                    context, {'id': user})
+            for user_package in user_packages:
+                if (user_package['package_id'] == package.id
+                        and user_package['capacity'] in ('admin', 'editor')):
+                    success = True
+                    break
```

Comment:
Query the table directly instead of doing this check in python so that users with large collaborator lists don't slow down package_update calls.

Reply:
I think I addressed this in 7d270b739